File: orig_app/internal/handler/security/web_ide_csp.go
package security

import (
	"net/url"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/jmadden/gitlab-redux/internal/config"
)

// WebIdeCSPHandler handles Content Security Policy for Web IDE
type WebIdeCSPHandler struct {
	config *config.Config
}

// NewWebIdeCSPHandler creates a new Web IDE CSP handler
func NewWebIdeCSPHandler(config *config.Config) *WebIdeCSPHandler {
	return &WebIdeCSPHandler{
		config: config,
	}
}

// IncludeWebIdeCSP adds Web IDE CSP directives to the response
func (h *WebIdeCSPHandler) IncludeWebIdeCSP(c *gin.Context) {
	// Get existing CSP directives
	csp := c.GetHeader("Content-Security-Policy")
	if csp == "" {
		return
	}

	// Parse the base URL
	baseURL, err := url.Parse(c.Request.URL.String())
	if err != nil {
		return
	}

	// Set the base path
	baseURL.Path = h.config.GitLab.RelativeURLRoot
	if baseURL.Path == "" {
		baseURL.Path = "/"
	}

	// Add webpack path
	baseURL.Path = strings.TrimRight(baseURL.Path, "/") + "/assets/webpack/"
	baseURL.RawQuery = "" // Remove query parameters

	// Get default-src directive
	defaultSrc := h.getDirectiveValues(csp, "default-src")
	if defaultSrc == "" {
		return
	}

	// Update frame-src directive
	frameSrc := h.getDirectiveValues(csp, "frame-src")
	if frameSrc == "" {
		frameSrc = defaultSrc
	}
	frameSrc += " " + baseURL.String() + " https://*.web-ide.gitlab-static.net/"
	csp = h.updateDirective(csp, "frame-src", frameSrc)

	// Update worker-src directive
	workerSrc := h.getDirectiveValues(csp, "worker-src")
	if workerSrc == "" {
		workerSrc = defaultSrc
	}
	workerSrc += " " + baseURL.String()
	csp = h.updateDirective(csp, "worker-src", workerSrc)

	// Set the updated CSP header
	c.Header("Content-Security-Policy", csp)
}

// getDirectiveValues extracts values from a CSP directive
func (h *WebIdeCSPHandler) getDirectiveValues(csp, directive string) string {
	parts := strings.Split(csp, ";")
	for _, part := range parts {
		part = strings.TrimSpace(part)
		if strings.HasPrefix(part, directive+" ") {
			return strings.TrimPrefix(part, directive+" ")
		}
	}
	return ""
}

// updateDirective updates a CSP directive with new values
func (h *WebIdeCSPHandler) updateDirective(csp, directive, values string) string {
	parts := strings.Split(csp, ";")
	for i, part := range parts {
		part = strings.TrimSpace(part)
		if strings.HasPrefix(part, directive+" ") {
			parts[i] = directive + " " + values
			return strings.Join(parts, "; ")
		}
	}
	return csp + "; " + directive + " " + values
}

File: src/controllers/concerns/web_ide_csp.rs
use actix_web::{web, HttpRequest, HttpResponse};
use std::collections::HashMap;
use url::Url;

pub trait WebIdeCSP {
    fn include_web_ide_csp(&self, req: &HttpRequest) -> Result<(), Box<dyn std::error::Error>>;
}

pub struct WebIdeCSPHandler;

impl WebIdeCSPHandler {
    pub fn new() -> Self {
        WebIdeCSPHandler
    }
}

impl WebIdeCSP for WebIdeCSPHandler {
    fn include_web_ide_csp(&self, req: &HttpRequest) -> Result<(), Box<dyn std::error::Error>> {
        // Get the current CSP directives from the request
        let mut csp_directives = req
            .headers()
            .get("content-security-policy")
            .and_then(|h| h.to_str().ok())
            .map(|h| parse_csp_directives(h))
            .unwrap_or_default();

        if csp_directives.is_empty() {
            return Ok(());
        }

        // Build the webpack URL
        let base_uri = req.uri().to_string();
        let mut url = Url::parse(&base_uri)?;
        let relative_url_root =
            std::env::var("GITLAB_RELATIVE_URL_ROOT").unwrap_or_else(|_| "/".to_string());
        url.set_path(&format!(
            "{}/assets/webpack/",
            relative_url_root.trim_end_matches('/')
        ));
        url.set_query(None);
        let webpack_url = url.to_string();

        // Update frame-src directive
        let default_src = csp_directives
            .get("default-src")
            .cloned()
            .unwrap_or_default();

        let frame_src = csp_directives
            .entry("frame-src".to_string())
            .or_insert_with(Vec::new);

        frame_src.extend(default_src.iter().cloned());
        frame_src.push(webpack_url.clone());
        frame_src.push("https://*.web-ide.gitlab-static.net/".to_string());

        // Update worker-src directive
        let worker_src = csp_directives
            .entry("worker-src".to_string())
            .or_insert_with(Vec::new);

        worker_src.extend(default_src);
        worker_src.push(webpack_url);

        // Set the updated CSP header
        let csp_header = format_csp_directives(&csp_directives);
        req.headers_mut()
            .insert("content-security-policy", csp_header.parse()?);

        Ok(())
    }
}

fn parse_csp_directives(header: &str) -> HashMap<String, Vec<String>> {
    let mut directives = HashMap::new();

    for directive in header.split(';') {
        if let Some((name, values)) = directive.split_once(' ') {
            let values: Vec<String> = values.split_whitespace().map(|s| s.to_string()).collect();
            directives.insert(name.to_string(), values);
        }
    }

    directives
}

fn format_csp_directives(directives: &HashMap<String, Vec<String>>) -> String {
    directives
        .iter()
        .map(|(name, values)| format!("{} {}", name, values.join(" ")))
        .collect::<Vec<_>>()
        .join("; ")
}
